Support for different types of bounding volumes for 3D tiles

e-andersson opened the pull request:
Bounding volume is now specified as `boundingVolume.box` or `boundingVolume.sphere` where it was previously only `box`. `boundingVolume.box` is the same as the old `box` for now.

PR for issue [Different bounding volumes](https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/10).

Somewhat rudimentary tests are included, feedback welcome.

e-andersson commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -187,10 +212,8 @@ define([
-        this._debugBox = undefined;
-        this._debugcontentBox = undefined;
-        this._debugOrientedBoundingBox = undefined;
-        this._debugContentsOrientedBoundingBox = undefined;
```
I kept only one debug volume for the tile and one debug volume for the tile contents. Not sure whether there is any need to keep them separated.

lilleyse replied:
This seems like the right approach to me. We will have to modify `Cities.html` to remove some of the extra buttons, but that can be a later commit.

e-andersson commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -295,66 +323,62 @@ define([
-        return createDebugPrimitive(geometry, color, Matrix4.fromRotationTranslation(obb.halfAxes, obb.center));
+    function createDebugVolume(boundingVolume, color) {
+        var geometry;
+        var modelMatrix = new Matrix4(
```
Will have this changed, doesn't need to be an identity matrix.

pjcozzi commented:
@lilleyse can you do the initial review?

@e-andersson we'll want to use `box` for an AABB.  What is a good name for the current box?  `region`?

e-andersson commented:
@pjcozzi Could we call it something along the lines of `curved_box` or perhaps `WGS84_box` (unless other coordinates may be used)? A little bit more verbose than `region` but possible clearer?

pjcozzi commented:
A lot of KML use cases will be served by served by 3D Tiles.  In particular, [KML regions](https://developers.google.com/kml/documentation/regions?hl=en), which call this `LatLonAltBox`.  So "region" will be familiar to many users (I would not use `LatLonAltBox`, by Cesium conventions, it would be `longitudeLatitudeHeightBox`).

By the way, we use camel case for property names so `curved_box` would be `curvedBox`.  There is an (incomplete) Spec Writing Guide here: https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/26

@lilleyse is there a more precise graphics term for this that we are over looking?

lilleyse commented:
I'm not sure what the proper term is. Maybe `tileBox` if that isn't too general.

lilleyse commented on Source/Core/BoundingSphere.js:
```diff
@@ -1109,6 +1109,10 @@ define([
+    BoundingSphere.distanceToCamera = function(sphere, frameState) {
+        return Math.max(0.0, Cartesian3.distance(sphere.center, frameState.camera.positionWC) - sphere.radius);
+    };
+
```
Add a comment above this function, but add `@private` at the end for now.

lilleyse replied:
Throw `DeveloperError` if `sphere` or `frameState` is not defined, similar to what the other functions do.

lilleyse commented on Source/Core/BoundingSphere.js:
```diff
@@ -1185,5 +1189,10 @@ define([
+    BoundingSphere.prototype.distanceToCamera = function(frameState) {
+        return BoundingSphere.distanceToCamera(this, frameState);
+    };
+
```
Add a comment here too.

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -212,7 +235,9 @@ define([
          * @readonly
          */
         orientedBoundingBox : {
+            //TODO deal with this property
```
It's probably best to remove this getter. I added it as a way to get the bounding volume for `Instanced` and `Points` content providers. I'll have a small commit handling this soon.

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -255,25 +280,28 @@ define([
     Cesium3DTile.prototype.visibility = function(cullingVolume) {
-        return cullingVolume.computeVisibilityWithPlaneMask(this._orientedBoundingBox, this.parentPlaneMask);
+        var boundingVolume = this._orientedBoundingBox || this._tileBoundingSphere;
+        return cullingVolume.computeVisibilityWithPlaneMask(boundingVolume, this.parentPlaneMask);
```
Change to `defaultValue(this._orientedBoundingBox, this._tileBoundingSphere)`

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -255,25 +280,28 @@ define([
      */
     Cesium3DTile.prototype.contentsVisibility = function(cullingVolume) {
-        if (!defined(this._contentsOrientedBoundingBox)) {
+        var boundingVolume = this._contentsOrientedBoundingBox || this._contentsBoundingSphere;
```
Same as above.

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -255,25 +280,28 @@ define([
      */
     Cesium3DTile.prototype.distanceToTile = function(frameState) {
-        return this._tileBoundingBox.distanceToCamera(frameState);
+        var boundingVolume = this._tileBoundingBox || this._tileBoundingSphere;
```
Same as above.

e-andersson commented:
@lilleyse `box` and `tileBox` may be a bit too similar, it's easier to see the difference between `box` and `region` (at least if one is familiar with regions from KML).

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -295,66 +323,62 @@ define([
+            0, 1, 0, 0,
+            0, 0, 1, 0,
+            0, 0, 0, 1
+        );
```
Cleaner to do `var modelMatrix = Matrix4.clone(Matrix4.IDENTITY);`

e-andersson replied:
Thanks -- I stumbled on the immutability of Matrix4.IDENTITY, didn't figure out I should clone it :)

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -295,66 +323,62 @@ define([
+            0, 0, 1, 0,
+            0, 0, 0, 1
+        );
+        if (boundingVolume.box) {
```
Use `defined`

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -295,66 +323,62 @@ define([
+                height : box[4],
+                extrudedHeight: box[5]
+             });
+        } else if (boundingVolume.sphere) {
```
Use `defined`

lilleyse commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -295,66 +323,62 @@ define([
-        if (owner.debugShowBoundingVolume) {
-            if (!defined(tile._debugOrientedBoundingBox)) {
-                tile._debugOrientedBoundingBox = createDebugOrientedBox(tile._orientedBoundingBox, hascontentBox ? Color.WHITE : Color.RED);
+        var showVolume = owner.debugShowBoundingVolume || (owner.debugShowContentBoundingVolume && !hasContentBoundingVolume);
```
Slight typo since the name in `Cesium3DTileset` is actually `debugShowContentsBoundingVolume` (extra `s`). It's probably best to edit the variable name in `Cesium3DTileset` and any occurrences in `Cities.html`.

e-andersson replied:
Do you mean we should leave it as it is here, and adjust to `debugShowContentBoundingVolume` in `Cesium3DTileset` and `Cities.html` instead of the original `debugShowContentsBoundingVolume`?
I figured I'd call it `...Content...` without the `s` since the tiles.json specification says `content` without the `s`.

lilleyse replied:
Yeah leave it here and change in the other files.

e-andersson replied:
Done!

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+             boundingVolume : {
+                 "sphere": [0.0, 0.0, 0.0, 5.0]
+             }
+        };
```
Cleaner to remove the quotation marks around the keys.

lilleyse replied:
Some formatting issues here with spacing.

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+    describe('bounding volumes', function() {
+        var tileWithBoundingSphere = {
+            "geometricError" : 1,
+            "refine" : "replace",
```
Use single quotes in general `"replace"` to `'replace'`

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+                 "children" : [],
+                 boundingVolume: {
+                    "box" : [rectangle[0], rectangle[1], rectangle[2], rectangle[3], -30, -34]
+                 }
```
Same formatting issues here.

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+    "use strict";
+
+    describe('bounding volumes', function() {
+        var tileWithBoundingSphere = {
```
I would just put all of sample headers like this at the top, outside of `describe` functions and share them among all the tests.

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+        it('can be a bounding box', function() {
+            var context = createContext();
+            var frameState = createFrameState(context, undefined, undefined, undefined);
+            var fakeTileset = {
```
Some earlier tests use the word 'fake', but use 'mock' instead.

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+        };
+
+        it('can be a bounding box', function() {
+            var context = createContext();
```
Easier to use `createScene()` instead. Check out [Instanced3DModel3DTileContentProviderSpec](https://github.com/AnalyticalGraphicsInc/cesium/blob/3d-tiles/Specs/Scene/Instanced3DModel3DTileContentProviderSpec.js) for an example. For update functions you can pass in `scene.frameState`.

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+
+        it('can have a bounding sphere', function() {
+            var tile = new Cesium3DTile(undefined, '/some_url', tileWithBoundingSphere, undefined);
+            expect(defined(tile._tileBoundingSphere)).toEqual(true);
```
Change to `expect(tile._tileBoundingSphere).toBeDefined()`

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+            var tile = new Cesium3DTile(undefined, '/some_url', tileWithContentsBoundingSphere, undefined);
+            expect(defined(tile._contentsBoundingSphere)).toEqual(true);
+            expect(tile._contentsBoundingSphere.radius).toEqual(5.0);
+            expect(tile._contentsBoundingSphere.center.equals(Cartesian3.ZERO)).toEqual(true);
```
Change to `expect(tile._contentsBoundingSphere.center).toEqual(Cartesian3.ZERO);`

lilleyse replied:
Same comment for rectangles.

lilleyse commented on Specs/Scene/Cesium3DTileSpec.js:
```diff
@@ -0,0 +1,149 @@
+            var tile = new Cesium3DTile(fakeTileset, '/some_url', tileWithBoundingBox, undefined);
+            tile.update(fakeTileset, frameState);
+            expect(defined(tile._debugBoundingVolume)).toEqual(true);
+            expect(tile._debugBoundingVolume._boundingSpheres.length).not.toEqual(0);
```
What is this checking?

e-andersson replied:
I wanted some check of whether the `_debugBoundingVolume` actually contains the right bounding sphere, but didn't quite get there. I'll have these checks removed.

lilleyse commented:
Those are all my comments. I did a quick test and it works fine.  When everything here is final, I'll update the test data sets to reflect the new `tiles.json` changes, as well as add some sphere and AABB tests (once we include AABB support).

When we pick a name for `region`, we may want to change the name of `TileBoundingBox` as well.

e-andersson commented:
Thanks @lilleyse. I'll update the code in line with the comments tomorrow morning.

lilleyse commented:
I added a `boundingVolume` getter to replace `orientedBoundingBox`. The getter will have to change once we add new box types.

pjcozzi commented:
Thanks for the review, @lilleyse!

Let me know when this is ready for me to look at.

e-andersson commented:
@lilleyse @pjcozzi I think this is ready now.

e-andersson commented:
Travis appears to disagree about the commit that dropped an `s` in `Contents`, but that shouldn't result in a build error, should it? There's no clear hint in the log.

e-andersson commented:
I updated the sample data, so the tests pass.

pjcozzi commented on Source/Core/BoundingSphere.js:
```diff
@@ -1110,6 +1110,26 @@ define([
+     *
+     * @private
+     */
+    BoundingSphere.distanceToCamera = function(sphere, frameState) {
```
Since `BoundingSphere` is in `Core`, and `FrameState` is in `Scene`, this really shouldn't take a `frameState` parameter.  Perhaps we can get around this by introducing a small private `BoundingSphere` wrapper in `Scene` that adds this function.

lilleyse replied:
`TileBoundingBox.distanceToCamera` also takes `frameState`.

pjcozzi replied:
Yes, but that is in `Scene`.

e-andersson replied:
I'll make a `TileBoundingSphere` in `Scene`.

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -206,14 +231,17 @@ define([
          * @readonly
          */
-        orientedBoundingBox : {
+        boundingVolume : {
```
Instead of returning `Object`, we should probably introduce an interface, `TileBoundingVolume` (in `Scene` for now).  For an example interface, see [ImageryProvider](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/ImageryProvider.js).

pjcozzi replied:
We should also rename this to `contentsBoundingVolume`, right?

e-andersson replied:
I don't know what this is actually used for, but it currently returns either the tile content bounding volume or the tile bounding volume. Should it be called `contentsBoundingVolume` even if it may return the tile's bounding volume?

lilleyse replied:
It's used [here](https://github.com/e-andersson/cesium/blob/3d-tiles_bounding_volumes/Source/Scene/Instanced3DModel3DTileContentProvider.js#L49) (which is later passed on to `ModelInstanceCollection`) and [here](https://github.com/e-andersson/cesium/blob/3d-tiles_bounding_volumes/Source/Scene/Points3DTileContentProvider.js#L58-L59). I think the name change is fine even if it returns the tile's bounding volume.

e-andersson replied:
So it seems [this](https://github.com/e-andersson/cesium/blob/3d-tiles_bounding_volumes/Source/Scene/Points3DTileContentProvider.js#L58-L59) might be a reason that some of the tests for `Points3DTileContentProvider` don't pass, right?

lilleyse replied:
Yeah sounds like it.

e-andersson replied:
Do you have any suggestions for how to get around that? Checking `instanceof`, and if the bounding volume is not a sphere, create one from the bounding box?

lilleyse replied:
That's probably the best approach for now. The code I was using before for the conversion is:

`this.boundingSphere = BoundingSphere.fromOrientedBoundingBox(tile.orientedBoundingBox);`

pjcozzi replied:
For the original name question, when the content bounding volume is not defined, it defaults to the tile's bounding volume, so this is the content's bounding volume even if it returns the tile's bounding volume.

e-andersson replied:
Changed the name from `boundingVolume` to `contentsBoundingVolume` in https://github.com/AnalyticalGraphicsInc/cesium/commit/f8a628e742cf0ef60559d505c6de154c6853aa75.

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -255,25 +283,28 @@ define([
      */
     Cesium3DTile.prototype.visibility = function(cullingVolume) {
-        return cullingVolume.computeVisibilityWithPlaneMask(this._orientedBoundingBox, this.parentPlaneMask);
+        var boundingVolume = defaultValue(this._orientedBoundingBox, this._tileBoundingSphere);
```
Instead of doing calls like this every frame:
```javascript
defaultValue(this._orientedBoundingBox, this._tileBoundingSphere)
```
Couldn't we just assign to a new `this._boundingVolume` in the constructor.  We might not even need to keep the "less preferred" bounding volumes in memory.

pjcozzi commented:
Thanks @e-andersson!  This is a great start.  Are you also up for adding OBB support in this PR?

e-andersson commented:
@pjcozzi Moved the distanceToCamera functionality to a new TileBoundingSphere. Will address the other issues tomorrow.
Perhaps the interface is best to leave for another PR? If we make an interface, currently only TileBoundingBox and TileBoundingSphere will adhere to it, right?

pjcozzi commented:
> If we make an interface, currently only TileBoundingBox and TileBoundingSphere will adhere to it, right?

Yes.

> Perhaps the interface is best to leave for another PR?

I think it is a small effort mainly for doc and consistency.  If we do this plus OBB, we'll be pretty confident in this part of the spec and Cesium implementation, and will hopefully not have to revisit it much later.

e-andersson commented:
Ok, I'll add the interface and OBB support in this PR. 

Currently Cesium3DTile uses a TileBoundingBox (if available) for the distanceToTile calculation, and otherwise an OrientedBoundingBox (if available), for example for visibility and contentsVisibility. Is this for performance reasons? If we don't need this, we could just assign to a `_boundingVolume` in the constructor and avoid keeping the different types of volumes around.

e-andersson commented:
Not quite sure about the interface. Should it expose properties, even if those properties are only available for one type of bounding volume? I don't think any properties actually need to be exposed, but doing it makes testing easier.

e-andersson commented:
@pjcozzi @lilleyse Please review the last commits. I've added a brief interface specification and support for OBBs (still using the boundingVolume.box property from tiles.json). Also see questions above.

lilleyse commented:
> Currently Cesium3DTile uses a TileBoundingBox (if available) for the distanceToTile calculation, and otherwise an OrientedBoundingBox (if available), for example for visibility and contentsVisibility. Is this for performance reasons? If we don't need this, we could just assign to a _boundingVolume in the constructor and avoid keeping the different types of volumes around.

I've been thinking about this too, keeping `TileBoundingBox` around just for the `distanceToTile` check seems unnecessary unless there is a performance reason. @pjcozzi do you have an idea about this?

lilleyse commented:
I made some formatting and cleanup changes. In the new bounding volume classes I removed some of the `DeveloperError` checks and default value settings because they are handled in the bounding volumes they call.

lilleyse commented on Source/Scene/TileBoundingVolume.js:
```diff
@@ -0,0 +1,43 @@
+     */
+    TileBoundingVolume.prototype.distanceToCamera = function(frameState) {
+        DeveloperError.throwInstantiationError();
+    };
```
I changed the instantiation errors to follow the convention in `Cesium3DTileContentProvider`, but I realize most of Cesium code base does it the other way.

lilleyse commented:
I added a getter to the `TileBoundingVolume` interface that returns the underlying bounding volume (sphere, obb, etc), which is important to have for instanced and point tiles. However right now it looks like `tile.boundingVolume.boundingVolume`, so I'm open to a better name. These changes fix all the broken tests.

I also fixed `Points3DTileContentProvider` so it converts to a bounding sphere if the tile's bounding volume is an OBB.

pjcozzi commented:
> Currently `Cesium3DTile` uses a `TileBoundingBox` (if available) for the `distanceToTile` calculation, and otherwise an `OrientedBoundingBox` (if available), for example for `visibility` and `contentsVisibility`. Is this for performance reasons? If we don't need this, we could just assign to a `_boundingVolume` in the constructor and avoid keeping the different types of volumes around. 

I _think_ I did this because we already had the `distanceToTile` for the `TileBoundingBox`, but did not have it for the `OrientedBoundingBox`.  When all bounding volumes implement this, we won't need both bounding volumes.

Likewise, `TileBoundingBox` will need to support the visibility test.  I suspect that doing an exact visibility test is slow and painful since, in the general case, this shape is not convex.  So I imagine it will store a conservative OBB to use for visibility.

I haven't looked at the latest code yet so you may have already figured all this out.

pjcozzi commented:
> Not quite sure about the interface. Should it expose properties, even if those properties are only available for one type of bounding volume? I don't think any properties actually need to be exposed, but doing it makes testing easier.

_Sometimes_ we expose `@private` properties like this to aid in writing better tests.

What is the specific case?

pjcozzi commented:
> I added a getter to the TileBoundingVolume interface that returns the underlying bounding volume (sphere, obb, etc), which is important to have for instanced and point tiles. However right now it looks like tile.boundingVolume.boundingVolume, so I'm open to a better name. These changes fix all the broken tests.

Just leave it or maybe `tile.boundingVolume.nativeBoundingVolume`.

pjcozzi commented on Source/Scene/TileBoundingVolume.js:
```diff
@@ -0,0 +1,53 @@
+     * @see TileBoundingBox
+     * @see TileBoundingSphere
+     */
+    var TileBoundingVolume = function() {
```
At quick glance, the interface looks good to me!

pjcozzi commented on Source/Scene/TileOrientedBoundingBox.js:
```diff
@@ -0,0 +1,65 @@
+            throw new DeveloperError('frameState is required.');
+        }
+        //>>includeEnd('debug');
+        return Math.sqrt(this.orientedBoundingBox.distanceSquaredTo(frameState.camera.positionWC));
```
@e-andersson did you test `distanceSquaredTo` to see if the numbers were reasonable?  There are unit tests for this, but it is not used anywhere else in Cesium so I just want to be sure.

e-andersson replied:
No, I haven't tested that. Will work a bit more on this today and see how it works.

pjcozzi commented:
This replaces `box` with an OBB, but to round out the spec and implementation, I think we need to include a `region` bounding volume.  Is that still the plan in this PR?

e-andersson commented:
@pjcozzi Yes, I'll have the `region` bounding volume added as well, just wanted some feedback on the work so far. I think I'll be done later today.

e-andersson commented:
>> Not quite sure about the interface. Should it expose properties, even if those properties are only available for one type of bounding volume? I don't think any properties actually need to be exposed, but doing it makes testing easier.

> Sometimes we expose @private properties like this to aid in writing better tests.

> What is the specific case?

`TileBoundingSphere` exposes `radius` which won't be a property of `TileOrientedBoundingBox`. Should `radius` then be listed as a property in `TileBoundingVolume`? `TileBoundingBox` also has a number of properties that are specific to it.

e-andersson commented:
@lilleyse 
> I added a getter to the TileBoundingVolume interface that returns the underlying bounding volume (sphere, obb, etc), which is important to have for instanced and point tiles. However right now it looks like tile.boundingVolume.boundingVolume, so I'm open to a better name. These changes fix all the broken tests.

> I also fixed Points3DTileContentProvider so it converts to a bounding sphere if the tile's bounding volume is an OBB.

`TileBoundingBox` doesn't have an underlying bounding volume so it just returns itself in my most recent (still local) version. Not sure what the needs of instanced and point tiles are, but this feels funny even if tests pass. As far as I can see there isn't any appropriate type in `core` that would be the underlying bounding volume used by `TileBoundingBox` -- should we add one for consistency?

e-andersson commented:
@pjcozzi 
> Likewise, TileBoundingBox will need to support the visibility test. I suspect that doing an exact visibility test is slow and painful since, in the general case, this shape is not convex. So I imagine it will store a conservative OBB to use for visibility.

> I haven't looked at the latest code yet so you may have already figured all this out.

I've implemented something like this now. `TileBoundingBox` creates and stores an OBB using two opposite corner points, and uses that OBB when checking for intersections. (I haven't tried finding out actual performance.)

lilleyse commented:
> TileBoundingBox doesn't have an underlying bounding volume so it just returns itself in my most recent (still local) version. Not sure what the needs of instanced and point tiles are, but this feels funny even if tests pass. As far as I can see there isn't any appropriate type in core that would be the underlying bounding volume used by TileBoundingBox -- should we add one for consistency?

It worked before because the bounding volume was always a `TileBoundingSphere` or `TileOrientedBoundingBox`, and the test data still uses 'box'. But yeah using `TileBoundingBox` will create problems.

I'm still not sure what the point of `TileBoundingBox` or the `region` type is right now.

lilleyse commented:
> TileBoundingSphere exposes radius which won't be a property of TileOrientedBoundingBox. Should radius then be listed as a property in TileBoundingVolume? TileBoundingBox also has a number of properties that are specific to it.

I think it's fine if each type has some unique properties like this.

pjcozzi commented:
> I'm still not sure what the point of TileBoundingBox or the region type is right now.

@lilleyse are you asking about the code?  Or the spec?

pjcozzi commented:
>> TileBoundingSphere exposes radius which won't be a property of TileOrientedBoundingBox. Should radius then be listed as a property in TileBoundingVolume? TileBoundingBox also has a number of properties that are specific to it.

> I think it's fine if each type has some unique properties like this.

:+1:

pjcozzi commented:
> I've implemented something like this now. TileBoundingBox creates and stores an OBB using two opposite corner points, and uses that OBB when checking for intersections. (I haven't tried finding out actual performance.)

I suspect performance is fine.

lilleyse commented:
> @lilleyse are you asking about the code? Or the spec?

Actually I take back what I said about the `region` type since it's useful to distinguish bounding volumes in Cartesian or Cartographic.

So mainly I'm talking about `TileBoundingBox` in the code, because it can use `OrientedBoundingBox` for most of its operations. However there may be uses for it that I haven't thought about.

pjcozzi commented:
@lilleyse `TileBoundingBox` (which we should rename to indicate `region`) should have a more precise distance check, which avoids refining too soon.  The top of an OBB is a plane; the top of a region is a curve that is tangent to the plane at one point with the rest being under the plane.

lilleyse commented:
Ah that makes sense.

e-andersson commented:
Shall we just call it `TileBoundingRegion` instead of `TileBoundingBox`?

pjcozzi commented:
> Shall we just call it TileBoundingRegion instead of TileBoundingBox?

:+1:

e-andersson commented:
Anything else that needs to be addressed in this PR? Better tests for the tile bounding volumes?

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -61,30 +67,69 @@ define([
-            rs = OrientedBoundingBox.fromRectangle(new Rectangle(cb[0], cb[1], cb[2], cb[3]), cb[4], cb[5]);
+            var headerVolume = contentHeader.boundingVolume;
+
+            if (defined(headerVolume.box)) {
```
This is basically duplicate with the `if` block above, right?  Can we introduce a helper that returns the bounding volume instance given the 3D Tiles `boundingVolume` property?

e-andersson replied:
Like [this](https://github.com/AnalyticalGraphicsInc/cesium/commit/2c52a01ed7da7e0ab18665db436d36be47fd593e)? I think it's a bit cleaner now.

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -295,66 +339,66 @@ define([
-            dimensions: new Cartesian3(2.0, 2.0, 2.0)
-        });
-        return createDebugPrimitive(geometry, color, Matrix4.fromRotationTranslation(obb.halfAxes, obb.center));
+    function createDebugVolume(boundingVolume, color) {
```
Is it reasonable to move the debug visualization into the new tile bounding volumes classes?

e-andersson replied:
Sounds reasonable to me.

pjcozzi replied:
Awesome, thanks!

pjcozzi commented on Source/Scene/TileBoundingRegion.js:
```diff
@@ -41,6 +45,16 @@ define([
+         * @type {OrientedBoundingBox}
+         * @default OrientedBoundingBox()
+         */
+        this.orientedBoundingBox = new OrientedBoundingBox();
```
Can this be private?  Just change it to:
```
this._orientedBoundingBox = new OrientedBoundingBox();
```
Also, I would just inline the `computeOrientedBoundingBox` function since it is awkward that we allocate the object here, then allocate it again.

pjcozzi commented:
@lilleyse any other comments?

I'll do the spec updates in the 3d-tiles repo today or tomorrow.

@e-andersson thanks again for this effort!

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -61,30 +67,69 @@ define([
-        });
-        this._orientedBoundingBox = OrientedBoundingBox.fromRectangle(rectangle, b[4], b[5]);
+        var tileBoundingVolume = header.boundingVolume;
+        if (defined(tileBoundingVolume.box)) {
```
I'm writing the spec now and I am a bit confused about `box`.  I thought the intention was for it to be an arbitrary object-oriented bounding box.  It would be an array of 12 numbers, with the first three being the center (x, y, z) and the next three being the columns for the half axes (x, y, z) similar to:

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/OrientedBoundingBox.js#L57

pjcozzi replied:
Here is what I have for the schema:
> An array of 12 numbers that define a oriented bounding box in WGS84 coordinates.  The first three elements define the x, y, and z values for the center of the box.  The next three elements (with indices 3, 4, and 5) define the x axis direction and length.  The next three elements (indices 6, 7, and 8) define the y axis direction and length.  The last three elements (indices 9, 10, and 11) define the z axis direction and length.

e-andersson replied:
@pjcozzi I went over that a bit too fast. I agree about `box` and have made some changes in https://github.com/AnalyticalGraphicsInc/cesium/commit/c7e9c9b23e99ba039e33efd274e90b90c983fbac, but I'm not confident that my usage of `halfAxes` in `OrientedBoundingBox` is right (see [Cesium3DTile](https://github.com/AnalyticalGraphicsInc/cesium/blob/c7e9c9b23e99ba039e33efd274e90b90c983fbac/Source/Scene/Cesium3DTile.js#L277-L285) and [TileOrientedBoundingBox](https://github.com/AnalyticalGraphicsInc/cesium/commit/c7e9c9b23e99ba039e33efd274e90b90c983fbac#diff-e9edfe6c30a08cde0c705cc817c945c5R18)). Does this look reasonable?

pjcozzi replied:
Looks good (just that one style comment).  If you want to test a simple case for visualization do something like:
```
box : [7000000, 0, 0,        200000, 0, 0,        0, 200000, 0,        0, 0, 200000]
```
That should make a square box centered at lat/lon (0, 0).

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -61,30 +67,69 @@ define([
         var contentHeader = header.content;
 
 // TODO: For the 3D Tiles spec, we want to change this to basically (x, y, z) to (x, y, z)
```
I think we can remove this `TODO`.  The current order is more intuitive and consistent with Cesium's `Rectangle`.

e-andersson replied:
Removed

pjcozzi commented:
Spec updates: https://github.com/AnalyticalGraphicsInc/3d-tiles/pull/51

pjcozzi commented:
> Like this? I think it's a bit cleaner now.

:+1: 

I also committed 91219cfb653a9b259045a43c268cca021be89a44 to ensure the JavaScript engine is able to create a hidden class (if the `if` statement was false, the property was never assigned to, and I'm not 100% sure JavaScript engines are smart enough to optimize this case).

See https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Documentation/Contributors/CodingGuide/README.md#constructor-functions

And this talk: https://www.youtube.com/watch?v=XAqIpGU8ZZk

e-andersson commented:
Visualization of TileOrientedBoundingBoxes suggest that something isn't quite right, the boxes jump around in an ominous way when the camera moves. Will have to test further.

e-andersson commented:
After visualizing `TileOrientedBoundingBox` more carefully in Sandcastle, it appears to work fine. Earlier I was probably confused by how the boxes were cutting through the ellipsoid (when there was no terrain or 3D tiles on it), in combination with poorly chosen positions/dimensions.
I also added a few tests for distance and plane intersection calculations. These are already tested much more thoroughly in `OrientedBoundingBoxSpec` but I thought it could be reassuring to have a few for `TileOrientedBoundingBox` as well to make sure it's properly wired up.

e-andersson commented:
@pjcozzi @lilleyse I think this is ready now.

e-andersson commented:
Almost ready, will have a look at some tests first.

e-andersson commented:
@lilleyse Would you mind having a look at why the `Instanced3DTileContentProvider` and `Composite3DTileContentProvider` resolve tests don't pass (timeout)? The sample data looks fine and there isn't any problem instantiating the bounding volumes, so I'm not sure where the problem is. Do you have an idea?

lilleyse commented:
It looks like the tests are failing because `ModelInstanceCollection` expects the bounding volume to have a `center` property in order to do relative to center rendering. So a quick fix would be to to add a `center` getter in `TileBoundingRegion`, or the `boundingVolume` getter could return `this._orientedBoundingBox` instead.

e-andersson commented:
@lilleyse Thanks, that indeed seems to be it.

e-andersson commented:
I added a temporary solution for creating an oriented bounding box from points representing the corners when creating a `TileBoundingRegion`. We'll probably not want when there is a `OrientedBoundingBox.fromCorners` method, but until then it's nice to have a real box.

e-andersson commented:
@pjcozzi  I think this is ready.

pjcozzi commented:
@e-andersson can you merge the `3d-tiles` branch into this?  Looks like there are some minor conflicts from @lilleyse's test branch that we merged today.

pjcozzi commented on Source/Scene/TileBoundingRegion.js:
```diff
@@ -99,8 +115,52 @@ define([
+         *
+         * @private
+         */
+        this._orientedBoundingBox = OrientedBoundingBox.fromPoints(getBoxPoints(this.rectangle, this.minimumHeight, this.maximumHeight));
```
I believe we can use `OrientedBoundingBox.fromRectangle` here.  For example, see https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Workers/createVerticesFromQuantizedTerrainMesh.js#L121

pjcozzi commented:
> @e-andersson can you merge the 3d-tiles branch into this? Looks like there are some minor conflicts from @lilleyse's test branch that we merged today.

Also, when you merge in master, please change the new tiles.json files for the tests in [Specs/Data/Cesium3DTiles](https://github.com/AnalyticalGraphicsInc/cesium/tree/3d-tiles/Specs/Data/Cesium3DTiles) so they follow the latest schema for bounding volumes

pjcozzi commented:
Just those comments.  Once the `3d-tiles` branch is merged in, I'll test this locally, and then we should be good to merge.

e-andersson commented:
Merged in 3d-tiles into this branch. Note that I seem to have introduced two test failures, possibly due to the `_orientedBoundingBox` in `TileBoundingRegion` being incorrectly created using `OrientedBoundingBox.fromPoints`. It seems this is also manifested as error-prone culling/tile selection in the viewer.
@pjcozzi and @lilleyse, does the above seem correct? If so, it will be fixed by introduction of a way to create oriented bounding boxes as discussed in https://github.com/AnalyticalGraphicsInc/cesium/pull/3325#commitcomment-15013420, right?

lilleyse commented:
Maybe you can try `OrientedBoundingBox.fromRectangle` from this comment https://github.com/AnalyticalGraphicsInc/cesium/pull/3325#discussion_r48283173

e-andersson commented:
Thanks for the pointer @lilleyse, I seem to have missed that comment. Works fine now.

This is ready now @pjcozzi.

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -426,10 +409,13 @@ define([
-        this._debugcontentBox = this._debugcontentBox && this._debugcontentBox.destroy();
-        this._debugOrientedBoundingBox = this._debugOrientedBoundingBox && this._debugOrientedBoundingBox.destroy();
-        this._debugContentsOrientedBoundingBox = this._debugContentsOrientedBoundingBox && this._debugContentsOrientedBoundingBox.destroy();
+        // TODO: Is there a reason to destroy the bounding volumes (and thus to implement destroy() functions for them)? The destroyObject function description says it should be used to for objects that have GL resources. 
```
This is only needed for the `debug` volumes since they contain Cesium primitives, which contain GL resources.

e-andersson replied:
Ah, got it. (Sean has fixed it.)

pjcozzi commented on Specs/Scene/Cesium3DTilesetSpec.js:
```diff
@@ -610,7 +610,7 @@ defineSuite([
         return Cesium3DTilesTester.loadTileset(scene, tilesetUrl).then(function(tileset) {
             viewRootOnly();
             // TODO : remove 's' when #3325 is merged in
```
Can we remove this `TODO`?

pjcozzi commented:
Thanks again @e-andersson!  I'm going to merge this since we have some changes we want to make that depend on this.  Can you please open a separate PR into the `3d-tiles` branch to address these two minor comments: https://github.com/AnalyticalGraphicsInc/cesium/pull/3325#discussion-diff-49740001 and https://github.com/AnalyticalGraphicsInc/cesium/pull/3325#discussion-diff-49739927

pjcozzi commented:
@tfili and anyone else using 3D Tiles - note that next time you update to the latest `3d-tiles` branch that this is a small breaking change.
